sprint_6-solution

A-Glebov opened the pull request:
1. В HistoryManager добавили  метод void remove(int id)
2. Добавили класс Node для реализации алгоритма со связанным списком
3. Реализовали алгоритм со связанным списком хранения истории просмотров в классе InMemoryHistoryManager и протестировали его
4. Реализовали пользовательский сценарий согласно ТЗ

A-Glebov committed:
1. В HistoryManager добавили  метод void remove(int id)
2. Добавили класс Node для реализации алгоритма со связанным списком
3. Реализовали алгоритм со связанным списком хранения истории просмотров в классе InMemoryHistoryManager и протестировали его
4. Реализовали пользовательский сценарий согласно ТЗ

BerezinYaroslav commented on src/service/HistoryManager.java:
```diff
@@ -5,7 +5,11 @@
 import java.util.List;
 
 public interface HistoryManager {
```
В интерфейсе все верно

BerezinYaroslav commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,28 +1,84 @@
+    Node last;
+
+    // Создание Node
+    void linkLast(Task task) {
```
⛔️ Нужно добавить модификатор доступа. Так как метод не должен использоваться извне - private

BerezinYaroslav replied:
Есть

BerezinYaroslav commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,28 +1,84 @@
+    }
+
+    // Удаление Node
+    void removeNode(Node node) {
```
⛔️ Нужно добавить модификатор доступа. ак как метод не должен использоваться извне - private

BerezinYaroslav replied:
Есть

BerezinYaroslav commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,28 +1,84 @@
+    }
+
+    // Получение списка задач из Node
+    public List<Task> getTasks() {
```
⛔️ Так как метод не должен использоваться извне - private

BerezinYaroslav replied:
Есть

BerezinYaroslav commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,28 +1,84 @@
-            history.removeFirst();
-        }
-        history.add(task);
+        remove(task.getId());
```
⛔️ Надо проверить это поле на null перед получением и использованием, иначе можем словить здесь ошибку

A-Glebov replied:
Добавил проверку на null 76 строка

BerezinYaroslav replied:
Ага, есть

BerezinYaroslav commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,28 +1,84 @@
+    public List<Task> getTasks() {
+        List<Task> history = new ArrayList<>();
+        Node currentNode = first;
+        while (currentNode != null) {
```
➕ While следует обособить отступами, в данном случае с обоих сторон, так как он не в начале или конце метода

BerezinYaroslav replied:
Так лучше, молодец)

BerezinYaroslav commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,28 +1,84 @@
 
     // Список для хранения истории просмотров
-    private final List<Task> history = new ArrayList<>();
+    Map<Integer, Node> nodes = new HashMap<>();
```
⛔️ Всем трем полям нужно добавить модификатор доступа

BerezinYaroslav replied:
Есть

BerezinYaroslav commented on src/Main.java:
```diff
@@ -11,14 +11,40 @@ public class Main {
 
-    // ТЕСТИРОВАНИЕ
-    private static TaskManager getInMemoryTaskManager() {
+    private static TaskManager getTaskManager() {
```
⛔️ Вроде понял твою логику, но такой подход сложен и не совсем верен (создать таск менеджер, "настроить" его и потом показать, что в нем содержится). Лучше все делать в метода main, без разделения на создание, "настройку" и отображение состояния. Если будет плохо смотреться, напиши мне, посоветую, как в данном случае лучше разделить на логические блоки

A-Glebov replied:
Перенес создание и настройку менеджера в майн. Отказаться ли от методов printHistory и printAllTAsks?

A-Glebov replied:
Методы остались с прошлых фз. Пусть будут так компактнее

BerezinYaroslav replied:
Пойдет

BerezinYaroslav reviewed (changes requested):
Алексей, привет еще раз)

Написал в коде комментарии, есть что поправить. Если будут вопросы - пиши мне в пачке

A-Glebov committed:
1. Проставили private модификаторы полям InMemoryHistoryManager
2. В методе remove добавили проверку на null
3. Изменили способ создания менеджера в Main

BerezinYaroslav commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -1,28 +1,89 @@
+    // Создание Node
+    private void linkLast(Task task) {
+        Node newNode = new Node(last, task, null);
+        if (first == null) {
```
➕ if- и swith- конструкции тоже следует обособлять отступами по тому же правилу, которое я писал для while

(написал только тут, но пробегись по всем файлам, проверь на этот момент)

BerezinYaroslav reviewed (approved):
Алексей, привет)

Все поправил, принято! Оставил одну маленькую рекомендацию, перед следующим ТЗ лучше внести ее

Удачи!
